core/src/main/scala/kafka/server/AdminManager.scala

```diff
@@ -115,9 +125,9 @@ class AdminManager(val config: KafkaConfig,
 
             // Use `null` for unset fields in the public API
             val numPartitions: java.lang.Integer =
-              if (topic.numPartitions == NO_NUM_PARTITIONS) null else topic.numPartitions
+              if (topic.numPartitions == NO_NUM_PARTITIONS) null else resolvedNumPartitions
             val replicationFactor: java.lang.Short =
-              if (topic.replicationFactor == NO_REPLICATION_FACTOR) null else topic.replicationFactor
+              if (topic.replicationFactor == NO_REPLICATION_FACTOR) null else resolvedReplicationFactor
```

Comment:
The documentation for `CreateTopicPolicy` says:

> @param numPartitions the number of partitions to create or null if replicasAssignments is set.
> @param replicationFactor the replication factor for the topic or null if replicaAssignments is set.

It seems like we may end up with a non null `numPartitions` and `replicationFactor` even if `replicaAssignment` is set?

Reply:
What's the scenario that causes this? If `replicaAssignments` is set, then we expect the partitions/replication factor to equal `NO_PARTITIONS` and `NO_REPLICATION_FACTOR` respectively, otherwise it fails here:
```java
        if ((topic.numPartitions != NO_NUM_PARTITIONS || topic.replicationFactor != NO_REPLICATION_FACTOR)
            && !topic.assignments().isEmpty) {
          throw new InvalidRequestException("Both numPartitions or replicationFactor and replicasAssignments were set. " +
            "Both cannot be used at the same time.")
        }
```

Reply:
You're right. But then why do we need this change at all? It seems like we never end up with a different result than the previous code.

Reply:
The difference is that if the value sent was `-1` (i.e. `NO_REPLICATION_FACTOR`) and there are no assignments, it will instead be set the `resolvedReplicationFactor`, which is the cluster default (i.e. `default.replication.factor`)

Reply:
Hmm.. I think Ismael's point is that if `topic.replicationFactor == NO_REPLICATION_FACTOR`, then `resolvedReplicationFactor == topic.replicationFactor`, and this is the equivalent to what we already had. It seems like this is what we're trying to do is something like this:
```scala
val replicationFactor: java.lang.Short =
  if (topic.assignments().nonEmpty) null else resolvedReplicationFactor
```

Reply:
🤦‍♂ I see what you're saying now. I think @hachikuji's code snippet is the easiest to read so I'll change it to that. @ijuma my original line of thought was that it's easier to reason about the code if we only ever use `resolvedX` everywhere, hence the motivation for this change. That way I would never accidentally use the value of `-1` in business logic code.